test/server/cards/01_SOR/events/Waylay.spec.js

```diff
@@ -0,0 +1,54 @@
+describe('Waylay', function() {
+    integration(function() {
+        describe('Waylay\'s ability', function() {
+            beforeEach(function () {
+                this.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['waylay'],
+                        groundArena: [{ card: 'pyke-sentinel', upgrades: ['entrenched'] }],
+                    },
+                    player2: {
+                        groundArena: ['wampa', 'superlaser-technician'],
+                        spaceArena: [{ card: 'imperial-interceptor', upgrades: ['academy-training'] }]
+                    }
+                });
+            });
+
+            it('can return a friendly or enemy unit to its owner\'s hand', function () {
+                this.player1.clickCard(this.waylay);
+                expect(this.player1).toBeAbleToSelectExactly([this.wampa, this.pykeSentinel, this.imperialInterceptor, this.superlaserTechnician]);
+
+                this.player1.clickCard(this.superlaserTechnician);
+                expect(this.superlaserTechnician).toBeInLocation('hand', this.player2);
+            });
+        });
+
+        describe('Waylay\'s ability', function() {
+            beforeEach(function () {
+                this.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['entrenched'],
+                    },
+                    player2: {
+                        hand: ['waylay'],
+                        groundArena: ['wampa'],
+                    }
+                });
+            });
+
+            it('when played on a friendly upgraded unit, it will cause the upgrade to be in the owner\'s discard', function () {
+                this.player1.clickCard(this.entrenched);
```

Comment:
Can you do me a small favor and add a comment here that says this:

`TODO: add an option to set an upgrade's controller in test setup`

We should make this simpler but it's not an urgent task